arrow-continuations/src/main/kotlin/arrow/continuations/generic/SuspendingComputation.kt

```diff
@@ -0,0 +1,112 @@
+package arrow.continuations.generic
+
+import kotlinx.atomicfu.atomic
+import kotlinx.atomicfu.loop
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlin.coroutines.intrinsics.COROUTINE_SUSPENDED
+import kotlin.coroutines.intrinsics.startCoroutineUninterceptedOrReturn
+import kotlin.coroutines.intrinsics.suspendCoroutineUninterceptedOrReturn
+import kotlin.coroutines.resumeWithException
+
+sealed class Either<out E, out A>
+data class Left<E>(val e: E) : Either<E, Nothing>()
+data class Right<A>(val a: A) : Either<Nothing, A>()
+
+interface EitherBind<E> {
+  suspend fun <A> Either<E, A>.invoke(): A
+}
+
+suspend fun <E, A> either(f: suspend EitherBind<E>.() -> A): Either<E, A> =
+  suspendCoroutineUninterceptedOrReturn { cont ->
+    SuspendMonadContinuation(cont).startCoroutineUninterceptedOrReturn {
+      f.invoke(object : EitherBind<E> {
+        override suspend fun <B> Either<E, B>.invoke(): B =
+          when(val ea = this) {
+            is Right -> ea.a
+            is Left -> shift(ea)
+          }
+      }).let(::Right)
+    }
+  }
+
+
+interface SuspendingComputation<R> {
+  suspend fun <A> shift(a: R): A
+}
+
+internal const val UNDECIDED = 0
+internal const val SUSPENDED = 1
+
+class ShortCircuit(val raiseValue: Any?) : ControlThrowable()
```

Comment:
If we move this to an inner class inside `SuspendingComputation` we can type this such as :
```
interface SuspendingComputation<R> {
  inner class ShortCircuit(val value: R) : ControlThrowable
  suspend fun <A> shift(a: R): A
}
```
This way we can ensure they can only short-circuit to valid values for that context

Reply:
Sounds good!

Reply:
Impossible: `Inner class of generic class extending 'Throwable' is prohibited`